File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Servo Bricklet 2.0

    ' Use position reached callback to swing back and forth
    Sub PositionReachedCB(ByVal sender As BrickletServoV2, _
                          ByVal servoChannel As Integer, ByVal position As Short)
        If position = 9000 Then
            Console.WriteLine("Position: 90°, going to -90°")
            sender.SetPosition(servoChannel, -9000)
        Else If position = -9000 Then
            Console.WriteLine("Position: -90°, going to 90°")
            sender.SetPosition(servoChannel, 9000)
        Else
            ' Can only happen if another program sets position
            Console.WriteLine("Error")
        End If
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim s As New BrickletServoV2(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register position reached callback to subroutine PositionReachedCB
        AddHandler s.PositionReachedCallback, AddressOf PositionReachedCB

        ' Enable position reached callback
        s.SetPositionReachedCallbackConfiguration(0, True)

        ' Set velocity to 100°/s. This has to be smaller or equal to the
        ' maximum velocity of the servo you are using, otherwise the position
        ' reached callback will be called too early
        s.SetMotionConfiguration(0, 10000, 500000, 500000)
        s.SetPosition(0, 9000)
        s.SetEnable(0, True)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()

        s.SetEnable(0, False)

        ipcon.Disconnect()
    End Sub
End Module

File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletServoV2;

type
  TExample = class
  private
    ipcon: TIPConnection;
    s: TBrickletServoV2;
  public
    procedure PositionReachedCB(sender: TBrickletServoV2; const servoChannel: word;
                                const position: smallint);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change XYZ to the UID of your Servo Bricklet 2.0 }

var
  e: TExample;

{ Use position reached callback to swing back and forth }
procedure TExample.PositionReachedCB(sender: TBrickletServoV2; const servoChannel: word;
                                     const position: smallint);
begin
  if (position = 9000) then begin
    WriteLn('Position: 90°, going to -90°');
    sender.SetPosition(servoChannel, -9000);
  end
  else if (position = -9000) then begin
    WriteLn('Position: -90°, going to 90°');
    sender.SetPosition(servoChannel, 9000);
  end
  else begin
    WriteLn('Error'); { Can only happen if another program sets position }
  end;
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  s := TBrickletServoV2.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register position reached callback to procedure PositionReachedCB }
  s.OnPositionReached := {$ifdef FPC}@{$endif}PositionReachedCB;

  { Enable position reached callback }
  s.SetPositionReachedCallbackConfiguration(0, true);

  { Set velocity to 100°/s. This has to be smaller or equal to the
    maximum velocity of the servo you are using, otherwise the position
    reached callback will be called too early }
  s.SetMotionConfiguration(0, 10000, 500000, 500000);
  s.SetPosition(0, 9000);
  s.SetEnable(0, true);

  WriteLn('Press key to exit');
  ReadLn;

  s.SetEnable(0, false);

  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.
